File: app/src/main/java/com/metahumanz/pacilread/importer/BookDuplicateDetector.java
package com.metahumanz.pacilread.importer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class BookDuplicateDetector {
    public enum MatchType { EXACT_CONTENT, SAME_TITLE_AUTHOR }

    public static final class Candidate {
        public final String key;
        public final String title;
        public final String author;
        public final String sha256;

        public Candidate(String key, String title, String author, String sha256) {
            this.key = key == null ? "" : key;
            this.title = title == null ? "" : title;
            this.author = author == null ? "" : author;
            this.sha256 = sha256 == null ? "" : sha256.trim().toLowerCase(Locale.ROOT);
        }
    }

    private BookDuplicateDetector() {
    }

    public static Map<String, MatchType> detect(List<Candidate> existing, List<Candidate> incoming) {
        Map<String, MatchType> result = new HashMap<>();
        List<Candidate> seen = new ArrayList<>();
        if (existing != null) seen.addAll(existing);
        if (incoming == null) return result;
        for (Candidate candidate : incoming) {
            MatchType match = findMatch(seen, candidate);
            if (match != null) result.put(candidate.key, match);
            seen.add(candidate);
        }
        return result;
    }

    private static MatchType findMatch(List<Candidate> candidates, Candidate target) {
        if (target == null) return null;
        if (!target.sha256.isEmpty()) {
            for (Candidate candidate : candidates) {
                if (target.sha256.equals(candidate.sha256)) return MatchType.EXACT_CONTENT;
            }
        }
        String targetIdentity = identity(target.title, target.author);
        if (targetIdentity.isEmpty()) return null;
        for (Candidate candidate : candidates) {
            if (targetIdentity.equals(identity(candidate.title, candidate.author))) {
                return MatchType.SAME_TITLE_AUTHOR;
            }
        }
        return null;
    }

    public static String identity(String title, String author) {
        String safeTitle = normalize(title);
        String safeAuthor = normalize(author);
        if (safeTitle.isEmpty() && safeAuthor.isEmpty()) return "";
        return safeTitle + "\u0000" + safeAuthor;
    }

    private static String normalize(String value) {
        if (value == null) return "";
        return value.trim().replaceAll("\\s+", " ").toLowerCase(Locale.ROOT);
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/importer/BookDuplicateDetector.kt
package com.metahumanz.pacilread.importer

import java.util.Locale

object BookDuplicateDetector {
    enum class MatchType { EXACT_CONTENT, SAME_TITLE_AUTHOR }

    class Candidate(key: String?, title: String?, author: String?, sha256: String?) {
        @JvmField val key: String = key ?: ""
        @JvmField val title: String = title ?: ""
        @JvmField val author: String = author ?: ""
        @JvmField val sha256: String = sha256?.trim()?.lowercase(Locale.ROOT) ?: ""
    }

    @JvmStatic
    fun detect(existing: List<Candidate?>?, incoming: List<Candidate?>?): Map<String, MatchType> {
        val result = HashMap<String, MatchType>()
        val seen = ArrayList<Candidate?>()
        if (existing != null) seen.addAll(existing)
        if (incoming == null) return result
        for (candidate in incoming) {
            val match = findMatch(seen, candidate)
            if (match != null) result[candidate!!.key] = match
            seen.add(candidate)
        }
        return result
    }

    private fun findMatch(candidates: List<Candidate?>, target: Candidate?): MatchType? {
        if (target == null) return null
        if (target.sha256.isNotEmpty()) {
            for (candidate in candidates) {
                if (target.sha256 == candidate!!.sha256) return MatchType.EXACT_CONTENT
            }
        }
        val targetIdentity = identity(target.title, target.author)
        if (targetIdentity.isEmpty()) return null
        for (candidate in candidates) {
            if (targetIdentity == identity(candidate!!.title, candidate.author)) return MatchType.SAME_TITLE_AUTHOR
        }
        return null
    }

    @JvmStatic
    fun identity(title: String?, author: String?): String {
        val safeTitle = normalize(title)
        val safeAuthor = normalize(author)
        if (safeTitle.isEmpty() && safeAuthor.isEmpty()) return ""
        return "$safeTitle\u0000$safeAuthor"
    }

    private fun normalize(value: String?): String =
        value?.trim()?.replace(Regex("\\s+"), " ")?.lowercase(Locale.ROOT) ?: ""
}
